### DISCUSSION TITLE
Ros2 Mapping Tutorial "IndexError: list assignment index out of range" (ubuntu 22)

### DISCUSSION THREAD

**Discussion by martincooney** [bitcraze/discussions#1662] [Q&A] [answered]
Hello, hope you are very well.
This is my first time posting here.
Recently we bought two crazyflies, one with the stem bundle sensors, which have been very fun to try so far.
Our first goal was to reproduce the tutorial below about mapping and navigation (and then go a bit beyond that, by adding some sensors and such), but I'm running into a problem right away that I don't know how to solve--with an error message "`IndexError: list assignment index out of range`"--and don't know how to go further. Do you have any idea how to solve the IndexError problem, or otherwise how we can get the tutorial to work, or any advice? (Details are below.)

-----------------------------------------------------------------------

Tutorial: https://www.bitcraze.io/2024/09/crazyflies-adventures-with-ros-2-and-gazebo/

Before doing the tutorial, I upgraded the firmware on the crazyradio and crazyflies, and was able to do the STEM ranging bundle tutorials (the crazyflie seems to move nicely on windows10 to do the "push" demo, and ubuntu 22 on rpi4 to do the "first flight" demo, and I also controlled it with the mobile app and a ps3 controller).

Starting the ros2 tutorial, I managed to get past a few challenges:
* `ros_gz_sim `seems not available for Humble on arm64; since I was trying first to use ubuntu 22 on a rpi4, I reinstalled all again on a separate laptop, which worked
* simple_mapper crashed, saying `bresenham module not found` (step 2 simulation) and bresenham, rowan, and nicegui not found (step 3 real world), which I got past by using: `pip3 install bresenham/rowan/nicegui`

After getting past the above, the simulation in step2 worked for me, but the problem appeared when trying step 3 "Simple mapping real world". When running the command `(ros2 launch crazyflie_ros2_multiranger_bringup simple_mapper_real.launch.py`), simple_mapper crashes after a few seconds with` "IndexError: list assignment index out of range [ERROR] [simple_mapper_multiranger-5] process has died"`. On rviz one can see lots of shapes that look a bit like match sticks appearing over time, as in the screenshots (it's like rviz thinks the robot might be moving even though it's still stationary). Just in case this didn't matter, I tried pressing "t" to get the crazyflie to take off, but this made the crazyflie fly diagonally into a wall and ram itself repeatedly (until finally stopping after a little while).

After this, I verified: the two stem bundle extension boards (ranger and flow) are firmly connected, the problem also occurs when the battery is fully charged and when I try to be very careful with not moving the crazyflie during its auto-calibration on start-up, the crazyflie still works perfectly for the "push" demo and hasn't broken, and the config file ‘c`razyflie_real_crazyswarm2.yaml`’ seems to have the proper address (it says `'radio://0/80/2M/E7E7E7E7E7'`, and for the push demo on win10 it is `'radio://0/80/2M'`). 
(Just to check the latter even further, I also installed the cfclient again, this time on the ubuntu laptop, which created a new problem with numpy versions (I had to downgrade from 1.26.4 to 1.23.5 to avoid an error with np.float and see the indexerror again), but the address shown in cfclient from scanning was also the same and I verified I could connect and the pose shown seemed to change reasonably when tilting the crazyflie.)
The output of ros2 node list as soon as the command is given is
```
$ ros2 node list
/crazyflie_server
/gui
/joy_node
/rviz2
/simple_mapper
/teleop
/transform_listener_impl_5ad92efacdb0
/vel_mux
```
Then with` IndexError, simple_mapper` vanishes.
(The room I'm trying is a medium-sized room, and I haven't tried a very small room out of fear of the drone crashing and getting broken. The problem seemed less bad when I put the drone up on a table once rather than on the floor, in that the map seemed a bit more stable, but it still occurred after a little while.)

Any advice would be appreciated! 
Thank you, best,

Martin

![Screenshot from 2024-11-21 14-00-37](https://github.com/user-attachments/assets/e5b62794-ccdb-46a5-a456-c4b7f9c3fce5)
![Screenshot from 2024-11-21 13-59-56](https://github.com/user-attachments/assets/c873c30e-896b-4258-96e0-879d6c36c610)

**Comment by knmcguire** [comment, marked as answer]
Hi! I had to update your post a bit since it was quite a lot of text and it was easy to miss the overview. The bullet points didn't had a space between the * and the word so it didn't render well. Also anything code related I've tried to put in `code blocks` (the '< >' button above) to make it more readable. J

As what I could digest from your issues, I'll try to split it up:

1.  **The index out of bounds:** There is currently no good protection once the robot's position goes outside of the maximum indicated map size. I've opened up an issue in the repository to address that: https://github.com/knmcguire/crazyflie_ros2_multiranger/issues/2. Currently the assumption is just that the map is big enough that this won't happen. Unfortunately that doesn't hold when 2 happens (see next point)
2. **The map and transform going crazy** This is due that the positioning with the flowdeck is not correct. During the robotics developer day I used the Lighthouse positioning system. The flowdeck will also work but it is more sensitive to environment conditions than the lighthouse is. So if the surface it is flying on is not ideal, or the flowdeck is too close to the ground. I've written a blogpost which also includes tips and tricks to get stable flight with the flowdeck that I recommend you to read: https://www.bitcraze.io/2023/11/go-with-the-flow-relative-positioning-with-the-flow-de
3. **ROS_gz_sim not working**. I'll update that in the tutorial that it needs to be on 64-bit x86. ROS2 humble on arm devices is not feature complete so not all packages are supported. Jazzy is though, but not sure how good gazebo works on that, but you could try it. I don't think Jazzy requires that many changes to the tutorial itself (famous last words).

I think I hit most of your issues, but let me know if I was missing any.

**Reply by martincooney** [reply]
Thank you for the fast/helpful response, this did the trick perfectly! :) I found a wooden floor outside my university, made the room as bright as possible, constructed a small mock-up environment out of cardboard, and was able to complete the tutorial. (A very small extra suggestion: if you plan to update the tutorial, it could be explicitly mentioned/repeated in step 4 that `export GZ_SIM_RESOURCE_PATH="/home/$USER/crazyflie_mapping_demo/simulation_ws/crazyflie-simulation/simulator_files/gazebo/" `  also has to have been called if it hasn't been already (just sourcing the ros2 packages via setup.bash initially gave me an error)).

**Reply by knmcguire** [reply]
Ah great!

and I've updated the tutorial with this info. Thanks for the suggestion.
